os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -156,59 +134,89 @@ static int binary_manager_load(int bin_idx)
 
-	bmvdbg("Open a directory, %s\n", BINARY_DIR_PATH);
+		bmvdbg("Open a directory, %s\n", BINARY_DIR_PATH);
 
-	/* Read each directory entry */
-	for (;;) {
-		struct dirent *entryp = readdir(dirp);
-		if (!entryp) {
-			/* Finished with this directory */
-			break;
-		}
-		/* Read and Verify a binary file */
-		if (DIRENT_ISFILE(entryp->d_type) \
-			&& !strncmp(entryp->d_name, bin_name, name_len) && entryp->d_name[name_len] == '_') {
-			snprintf(filepath[file_idx], CONFIG_PATH_MAX, "%s/%s", BINARY_DIR_PATH, entryp->d_name);
-			ret = binary_manager_read_header(filepath[file_idx], &header_data[file_idx]);
-			if (ret == OK) {
-				valid_bin_count++;
-				version = (int)atoi(header_data[file_idx].bin_ver);
-				bmvdbg("Found valid header in version %d\n", version);
-				if (version > latest_ver) {
-					latest_ver = version;
-					latest_idx = file_idx;
-					bmvdbg("Latest version %d, idx %d\n", latest_ver, latest_idx);
+		/* Read each directory entry */
+		for (;;) {
+			struct dirent *entryp = readdir(dirp);
+			if (!entryp) {
+				/* Finished with this directory */
+				break;
+			}
+			/* Read and Verify a binary file */
+			if (DIRENT_ISFILE(entryp->d_type) \
+				&& !strncmp(entryp->d_name, bin_name, name_len) && entryp->d_name[name_len] == '_') {
+				snprintf(filepath[file_idx], CONFIG_PATH_MAX, "%s/%s", BINARY_DIR_PATH, entryp->d_name);
+				ret = binary_manager_read_header(filepath[file_idx], &header_data[file_idx]);
+				if (ret == OK) {
+					valid_bin_count++;
+					version = (int)atoi(header_data[file_idx].bin_ver);
+					bmvdbg("Found valid header in version %d\n", version);
+					if (version > latest_ver) {
+						latest_ver = version;
+						latest_idx = file_idx;
+						bmvdbg("Latest version %d, idx %d\n", latest_ver, latest_idx);
+					}
 				}
+				file_idx++;
 			}
-			file_idx++;
 		}
-	}
-	closedir(dirp);
+		closedir(dirp);
 
-	if (valid_bin_count == 0) {
-		bmdbg("No valid binary in fs : %s\n", BIN_NAME(bin_idx));
-		return ERROR;
+		if (valid_bin_count == 0) {
+			bmdbg("No valid binary in fs : %s\n", BIN_NAME(bin_idx));
+			return ERROR;
+		}
 	}
-
 	do {
-		ret = binary_manager_load_binary(bin_idx, filepath[latest_idx], &header_data[latest_idx]);
-		if (ret == OK) {
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+		if (!binp)
+#endif
+		{
+			strncpy(load_attr.bin_name, header_data->bin_name, BIN_NAME_MAX);
+			load_attr.bin_size = header_data->bin_size;
+			load_attr.compression_type = header_data->compression_type;
+			load_attr.ram_size = header_data->bin_ramsize;
+			load_attr.stack_size = header_data->bin_stacksize;
+			load_attr.priority = header_data->bin_priority;
+			load_attr.offset = CHECKSUM_SIZE + header_data->header_size;
+			strncpy(load_attr.bin_ver, header_data->bin_ver, BIN_VER_MAX);
+		}
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+		else {
+			load_attr = BIN_LOAD_ATTR(bin_idx);
+		}
+		load_attr.binp = binp;
+#endif
+		ret = binary_manager_load_binary(bin_idx, filepath[latest_idx], &load_attr);
```

Comment:
What if latest_idx is -1?

Reply:
A variable, latest_idx can be greater than or equal with 0 here by checking valid_bin_count variable line 191.
This is because latest_idx is updated with valid_bin_count.

Reply:
@jeongarmy But there is a path that makes latest_idx -1 even though valid_bin_count is greater than zero. Please, take a careful look again.